### PR TITLE
[Dropbox] Add support for more formats and password protected files

### PR DESCRIPTION
**Pull request by zenerdi0de**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds support for downloading transcoded formats and support for password protected files.

### CODE REVIEW

File: yt_dlp/extractor/dropbox.py [hunk 1]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
...
+        json_string = self._html_search_regex(r'InitReact\.mountComponent.+ "props":(.+), "elem_id"', webpage, 'Info JSON')
+        info_json = self._parse_json(json_string, video_id)
+        transcode_url = traverse_obj(info_json, ((None, 'preview'), 'file', 'preview', 'content', 'transcode_url'), get_all=False)
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(transcode_url, video_id)
+
+        # downloads enabled we can get the original file
+        if 'anonymous' in (try_get(info_json, lambda x: x['sharePermission']['canDownloadRoles']) or []):
+            video_url = re.sub(r'[?&]dl=0', '', url)
+            video_url += ('?' if '?' not in video_url else '&') + 'dl=1'
+            formats.append({'url': video_url, 'format_id': 'original', 'format_note': 'Original'})
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:69]:
```suggestion
            formats.append({'url': video_url, 'format_id': 'original', 'format_note': 'Original', 'quality': 1})
        self._sort_formats(formats)
```

File: yt_dlp/extractor/dropbox.py [hunk 2]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
     ]
 
     def _real_extract(self, url):
+
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:34]:
```suggestion
```

File: yt_dlp/extractor/dropbox.py [hunk 3]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
...
         mobj = self._match_valid_url(url)
         video_id = mobj.group('id')
+        webpage = self._download_webpage(url, video_id)
         fn = compat_urllib_parse_unquote(url_basename(url))
         title = os.path.splitext(fn)[0]
-        video_url = re.sub(r'[?&]dl=0', '', url)
-        video_url += ('?' if '?' not in video_url else '&') + 'dl=1'
+
+        password = self.get_param('videopassword')
+        if self._og_search_title(webpage) == 'Dropbox - Password Required' or 'Enter the password for this link' in webpage:
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:42]:
```suggestion
        if (self._og_search_title(webpage) == 'Dropbox - Password Required'
                or 'Enter the password for this link' in webpage):
```

File: yt_dlp/extractor/dropbox.py [hunk 4]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
...
+
+            if password:
+                content_id = self._search_regex(r'''content_id=(.*?)[\"\']''', webpage, 'content_id')
+                payload = f'is_xhr=true&t={self._get_cookies("https://www.dropbox.com").get("t").value}&content_id={content_id}&password={password}&url={url}'
+                response = self._download_json('https://www.dropbox.com/sm/auth', video_id, 'POSTing video password', data=payload.encode('UTF-8'), headers={'content-type': 'application/x-www-form-urlencoded; charset=UTF-8'})
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:47]:
```suggestion
                response = self._download_json(
                    'https://www.dropbox.com/sm/auth', video_id, 'POSTing video password', data=payload.encode('UTF-8'),
                    headers={'content-type': 'application/x-www-form-urlencoded; charset=UTF-8'})
```
line too long

File: yt_dlp/extractor/dropbox.py [hunk 5]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
...
         fn = compat_urllib_parse_unquote(url_basename(url))
         title = os.path.splitext(fn)[0]
-        video_url = re.sub(r'[?&]dl=0', '', url)
-        video_url += ('?' if '?' not in video_url else '&') + 'dl=1'
+
+        password = self.get_param('videopassword')
+        if self._og_search_title(webpage) == 'Dropbox - Password Required' or 'Enter the password for this link' in webpage:
+
+            if password:
+                content_id = self._search_regex(r'''content_id=(.*?)[\"\']''', webpage, 'content_id')
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:45]:
```suggestion
                content_id = self._search_regex(r'content_id=(.*?)["\']', webpage, 'content_id')
```
unnecessary escapes

File: yt_dlp/extractor/dropbox.py [hunk 6]
```diff
@@ -26,15 +31,46 @@ class DropboxIE(InfoExtractor):
...
+                if response.get('status') == 'authed':
+                    webpage = self._download_webpage(url, video_id)
+                else:
+                    raise ExtractorError('Authentication failed!', expected=True)
+
+            elif self._get_cookies('https://dropbox.com').get('sm_auth'):
+                webpage = self._download_webpage(url, video_id)
+
+            else:
+                raise ExtractorError('Password protected video, use --video-password <password>')
```
Comment (pukkandan) [inline yt_dlp/extractor/dropbox.py:58]:
```suggestion
                if response.get('status') != 'authed':
                    raise ExtractorError('Authentication failed!', expected=True)
                webpage = self._download_webpage(url, video_id)

            elif self._get_cookies('https://dropbox.com').get('sm_auth'):
                webpage = self._download_webpage(url, video_id)
            else:
                raise ExtractorError('Password protected video, use --video-password <password>', expected=True)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by Shelfen** [comment]
Seems Dropbox is unable to download any files now.

ERROR: [Dropbox] 3hqzs30vjbtinvj: Unable to extract Info JSON; please report this issue on  https://github.com/yt-dlp/yt-dlp , filling out the "Broken site" issue template properly. Confirm you are on the latest version using -U; please report this issue on  https://github.com/yt-dlp/yt-dlp , filling out the "Broken site" issue template properly. Confirm you are on the latest version using -U

**Comment by ndyanx** [comment]
```
def _real_extract(self, url):
        mobj = self._match_valid_url(url)
        video_id = mobj.group('id')
        webpage = self._download_webpage(url, video_id)
        fn = urllib.parse.unquote(url_basename(url))
        title = os.path.splitext(fn)[0]
        password = self.get_param('videopassword')
        thumbnail = None

        for encoded in reversed(re.findall(r'registerStreamedPrefetch\s*\(\s*"[\w/+=]+"\s*,\s*"([\w/+=]+)"', webpage)):
            part = base64.b64decode(encoded).decode('utf-8', 'ignore')
            if 'sm/password' in part:
                webpage = self._download_webpage(
                    'https://www.dropbox.com/sm/password?'+ part.split('?')[1], video_id)

        if (self._og_search_title(webpage) == 'Dropbox - Password Required'
                or 'Enter the password for this link' in webpage):

            if password:
                content_id = self._search_regex(r'content_id=(.*?)["\']', webpage, 'content_id')
                payload = f'is_xhr=true&t={self._get_cookies("https://www.dropbox.com").get("t").value}&content_id={content_id}&password={password}&url={url.replace("https://www.dropbox.com", "")}'
                response = self._download_json(
                    'https://www.dropbox.com/sm/auth', video_id, 'POSTing video password', data=payload.encode(),
                    headers={'content-type': 'application/x-www-form-urlencoded; charset=UTF-8'})
                if response.get('status') != 'authed':
                    raise ExtractorError('Authentication failed!', expected=True)
                webpage = self._download_webpage(url, video_id)
            elif self._get_cookies('https://dropbox.com').get('sm_auth'):
                webpage = self._download_webpage(url, video_id)
            else:
                raise ExtractorError('Password protected video, use --video-password <password>', expected=True)

        formats, subtitles, has_anonymous_download = [], {}, False
        for encoded in reversed(re.findall(r'registerStreamedPrefetch\s*\(\s*"[\w/+=]+"\s*,\s*"([\w/+=]+)"', webpage)):
            decoded = base64.b64decode(encoded).decode('utf-8', 'ignore')
            if not has_anonymous_download:
                has_anonymous_download = self._search_regex(
                    r'(anonymous:\tanonymous)', decoded, 'anonymous', default=False)
            transcode_url = self._search_regex(
                r'\n.(https://[^\x03\x08\x12\n]+\.m3u8)', decoded, 'transcode url', default=None)
            if not transcode_url:
                continue
            formats, subtitles = self._extract_m3u8_formats_and_subtitles(transcode_url, video_id, 'mp4')
            thumbnail = self._search_regex(
                r'(https://www\.dropbox\.com/temp_thumb_from_token/c/(.*?)\?preserve_transparency=False&rlkey=(.*?)&secure_hash=&size=(.*?)&size_mode=4)', decoded, 'thumbnail url', default=None)
            break

        # downloads enabled we can get the original file
        if has_anonymous_download:
            formats.append({
                'url': update_url_query(url, {'dl': '1'}),
                'format_id': 'original',
                'format_note': 'Original',
                'quality': 1,
            })

        return {
            'id': video_id,
            'thumbnail': thumbnail,
            'title': title,
            'formats': formats,
            'subtitles': subtitles,
        }
``` 
support for password, thumbnail

**Comment by zenerdi0de** [comment]
Hi @ndyanx could you create a new PR for this?

**Comment by bashonly** [comment]
@ndyanx or If you don't want to open a PR yourself, could you re-post the patch in #9864 and also state that you release the patch's code into the public domain (for use in yt-dlp)?

**Pull request merged**